workspace/__lib__/xod/datetime/unpack-datetime/patch.cpp

```diff
@@ -0,0 +1,59 @@
+
+struct State {
+};
+
+// clang-format off
+{{ GENERATED_CODE }}
+// clang-format on
+
+bool isLeapYear(int y) {
+  return (1970 + y > 0) && !((1970 + y) % 4) && (((1970 + y) % 100) || !((1970 + y) % 400));
+}
+
+void evaluate(Context ctx) {
+    auto time = getValue<input_IN>(ctx);
+
+    static const uint8_t monthDays[] = { 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31 };
+
+    emitValue<output_SEC>(ctx, time % 60);
+    time /= 60;
+    emitValue<output_MIN>(ctx, time % 60);
+    time /= 60;
+    emitValue<output_HOUR>(ctx, time % 24);
+    time /= 24;
+
+    emitValue<output_WDAY>(ctx, ((time + 4) % 7) == 0 ? 7 : ((time + 4) % 7));
+
+    uint16_t year = 0;
+    uint32_t days = 0;
+
+    while ((unsigned)(days += (isLeapYear(year) ? 366 : 365)) <= time) {
+        year++;
```

Comment:
This one is tricky. Better use do/while

```cpp
do {
    days += isLeapYear(year) ? 366 : 365;
    ++year;
} while (days <= time);
```

Reply:
These circles have different results

Reply:
`    uint32_t days = isLeapYear(year) ? 366 : 365;

    do {
        days += isLeapYear(year) ? 366 : 365;
        ++year;
    } while (days <= time);
` 
It works fine like this. What variant is better?

Reply:
@nkrkv 
I think the 
```cpp
 while ((days += (isLeapYear(year) ? 366 : 365)) <= time) {
        year++;
    }
``` 
is shortest and the order of both incrementions is understandable